### PR TITLE
Core/Scripts Boss Skadi the Ruthless rework

### LINKED ISSUE #15362: [3.3.5] Boss/Script: Utgarde Pinnacle Skadi movements
If Grauf flying to the Sky it is more a Walking in the Sky
EDIT by @Killyana:
To do fix Skadi movements and maybe the waypoint during the flying phase

### PR DESCRIPTION
**Pull request by sirikfoll**
**Changes proposed**:
- Full rewrite.

**Target branch(es)**: 335/6x

**Issues addressed**: Closes #15362 #2660

**Tests performed**: Built and solo test

**Known issues and TODO list**:
- [x] Freezing Cloud could be improved, is currently damaging the mobs too, yet, I don't know any better way to do it.

### CODE REVIEW

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 1]
```diff
@@ -15,474 +15,931 @@
...
-                    if (m_uiPoisonedSpearTimer <= diff)
-                    {
-                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM))
-                            DoCast(target, SPELL_POISONED_SPEAR);
-                        m_uiPoisonedSpearTimer = 10000;
-                    } else m_uiPoisonedSpearTimer -= diff;
+            me->setActive(true);
+            me->SetCanFly(true);
+            me->SetDisableGravity(true);
+            me->SetUInt32Value(UNIT_FIELD_BYTES_1, 50331648);
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: what's this ?

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 2]
```diff
@@ -15,474 +15,931 @@
...
+    public:
+        spell_summon_gauntlet_mobs_periodic() : SpellScriptLoader("spell_summon_gauntlet_mobs_periodic") { }
+
+        class spell_summon_gauntlet_mobs_periodic_AuraScript : public AuraScript
+        {
+            PrepareAuraScript(spell_summon_gauntlet_mobs_periodic_AuraScript);
+
+            void CastTheNextTwoSpells()
+            {
+                uint32 spellId = 0;
```
Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: no need to declare spellId outside of the loop and initialize it, it looks weird. please declare it at like 850 where you are assigning it a value already.

**Comment by Aokromes** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp:500:29: fatal error: 
      unused parameter 'summoner' [-Wunused-parameter]
    void IsSummonedBy(Unit* summoner) override

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 3]
```diff
@@ -15,474 +15,931 @@
...
+                for (uint8 i = 0; i < 2; ++i)
+                {
+                    spellId = SummonSpellsList.front();
```
Comment (DDuarte) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: Multiple instances of this `AuraScript` will modify the same list. Isn't it an issue?

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp [hunk 4]
```diff
@@ -15,474 +15,931 @@
...
+};
+
+enum CombatPhase
+{
+    PHASE_FLYING                         = 0,
+    PHASE_GROUND
 };
 
-enum Creatures
+std::list<uint32> SummonSpellsList =
```
Comment (DDuarte) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: Prefer `vectors` to `lists` in most cases.

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp (hunk 1 above)
Comment (sirikfoll) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: not sure what it does, founded in sniff
```
[0] UpdateType: Values
[0] GUID: Full: 0xF15000690D20B983 Type: Vehicle Entry: 26893 (Grauf) Low: 2144643
[0] UNIT_FIELD_BYTES_1: 50331648/3.761582E-37

Comment (DDuarte) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: This is equivalent to `me->SetByteFlag(UNIT_FIELD_BYTES_1, UNIT_BYTES_1_OFFSET_ANIM_TIER, UNIT_BYTE1_FLAG_ALWAYS_STAND | UNIT_BYTE1_FLAG_HOVER);`

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp (hunk 4 above)
Comment (Riztazz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: Yeah, for people that wonder why http://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html ;p

Comment (sirikfoll) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: std::vector don't have .pop_front(), which I need, any suggestions on a replace for this one?

Comment (DDuarte) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: Use a queue then.

Also,
> Multiple instances of this AuraScript will modify the same list. Isn't it an issue?

Comment (sirikfoll) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: well, it shouldn't be an issue, this aura will be used one time and only for 1 npc.

Comment (DDuarte) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: 1 npc per active instance

Comment (jackpoz) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: > this aura will be used one time and only for 1 npc.

this aura will be used by anyone casting the spell, even from the chat commands, modifying a shared object from multiple threads and crashing the server.
global object must always be read-only or protected by a lock mechanism to ensure thread-safety.

Comment (sirikfoll) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: yeah, it's bad, but like i said, right now, i don't know a better way, any suggestions?

> Please don't complain without suggesting a better way to do it, this is the better I managed to do.

Comment (DDuarte) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: We might have different definitions of "complaining".

You can move the queue (currently list) to the AuraScript scope.

Comment (sirikfoll) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp]: not meant to be rude, blame my english '-'
I will change and tested everything tomorrow, thanks!

**Comment by sirikfoll** [comment]
Anyone want's to try fixing the Freezing Cloud spell, i don't know how to fix it, everything else works

**Comment by Silvertower** [comment]
Maybe adding something like:

`targets.remove_if(Trinity::ObjectTypeIdCheck(TYPEID_PLAYER, false));`

to spell_freezing_cloud_area_X's FilterTargets() method.

There's probably a better filter though, as I'm not sure if this applies to pets (or if it should). It's been years since I did this fight. Just an idea.

**Comment by sirikfoll** [comment]
Finished ^^

**Pull request merged**